state/git/impl.go

```diff
@@ -0,0 +1,193 @@
+// Copyright © 2017 The Kubicorn Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/kris-nova/kubicorn/apis/cluster"
+	"github.com/kris-nova/kubicorn/cutil/logger"
+	"github.com/kris-nova/kubicorn/state"
+
+	g "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/config"
+)
+
+type JSONGitStoreOptions struct {
+	BasePath     string
+	ClusterName  string
+	CommitConfig *JSONGitCommitConfig
+}
+
+type JSONGitCommitConfig struct {
+	Name   string
+	Email  string
+	Remote string
+}
+
+// JSONGitStore exists to save the cluster state as a git change.
+type JSONGitStore struct {
+	options      *JSONGitStoreOptions
+	ClusterName  string
+	BasePath     string
+	AbsolutePath string
+}
+
+func NewJSONGitStore(o *JSONGitStoreOptions) *JSONGitStore {
+	return &JSONGitStore{
+		options:      o,
+		ClusterName:  o.ClusterName,
+		BasePath:     o.BasePath,
+		AbsolutePath: filepath.Join(o.BasePath, o.ClusterName),
+	}
+}
+
+func (git *JSONGitStore) Exists() bool {
+	if _, err := os.Stat(git.AbsolutePath); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func (git *JSONGitStore) Write(relativePath string, data []byte) error {
+	fqn := filepath.Join(git.AbsolutePath, relativePath)
+	err := os.MkdirAll(path.Dir(fqn), 0700)
+	if err != nil {
+		return err
+	}
+	fo, err := os.Create(fqn)
+	if err != nil {
+		return err
+	}
+	defer fo.Close()
+
+	_, err = io.Copy(fo, strings.NewReader(string(data)))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (git *JSONGitStore) Read(relativePath string) ([]byte, error) {
+	fqn := filepath.Join(git.AbsolutePath, relativePath)
+	bytes, err := ioutil.ReadFile(fqn)
+
+	//if there's nothing check the remote repo
+	if err != nil {
+		// Create a new repository
+		r, _ := g.NewFilesystemRepository(fqn)
```

Comment:
`fqn` is full path to the cluster file. As we're expecting it to stay the file, we can't create filesystem there. I'm not sure what's exactly the `git.AbsolutePath` but I think it would probably suit it the best.